### PR TITLE
Snapping when moving the cursor with keyboard

### PR DESCRIPTION
**Pull request by crsib**
Resolves: #4556

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by petersampsonaudacity** [comment]
Testing on W10 with @crsib 's branch build for this fix:  audacity-win-3.3.0-alpha-20230417+90bbe68-x64

This issue #4556 appears to be properly fixed in Dmitry's branch build.

I tested:
a) left/right arrows  -  moves the cursor by the snapping format
b) Shift + left/right arrows  -  expands the selection by the snapping format
c) Ctrl + Shift + left/right arrows  -  shrinks the selection by the snapping format

File: libraries/lib-snapping/SnapUtils.cpp [hunk 1]
```diff
@@ -246,142 +250,100 @@ namespace
...
-      // DV: It was decided that for the time being,
-      // BPM sets the duration for quarter notes.
-      // For this reason, `cfg.timeSignature.second` is ignored
-      const auto fracDuration = quarterDuration * 4.0 / divisor;
-      const auto multiplier = 1.0 / fracDuration;
-
-      return SnapWithMultiplier(value, multiplier, nearest);
-   };
-}
+      if (!mMultiplierFunctor)
```
Comment (Paul-Licameli) [inline libraries/lib-snapping/SnapUtils.cpp:311]: Assertion in the constructor, and constancy of the member, should make the null check unnecessary

File: libraries/lib-snapping/SnapUtils.cpp [hunk 2]
```diff
@@ -246,142 +250,100 @@ namespace
...
-      "cd", XO("CD frames"), false,
-      SnapFunction("cd_75_fps", XO("CDDA frames (75 fps)"), SnapToFrames(75.0)))
-};
+Registry::BaseItemPtr TimeInvariantSnapFunction(
+   const Identifier& functionId, const TranslatableString& label,
+   MultiplierFunctor functor)
+{
+   return std::make_unique<ProjectDependentMultiplierSnapItem>(
+      functionId, label, std::move(functor));
+}
```
Comment (Paul-Licameli) [inline libraries/lib-snapping/SnapUtils.cpp:341]: My latest PR "namespace-Callable" will conflict, but once this is merged into master, I will rebase, and this bit of new annoyance code will also simplify

File: src/menus/SelectMenus.cpp [hunk 3]
```diff
@@ -166,26 +166,19 @@ void SeekWhenAudioActive(double seekStep, wxLongLong &lastSelectionAdjustment)
 double GridMove
 (AudacityProject &project, double t, int minPix)
 {
-   auto &formats = ProjectNumericFormats::Get(project);
```
Comment (Paul-Licameli) [inline src/menus/SelectMenus.cpp:169]: Good to see this; now, you can remove "ProjectNumericFormats.h" and "NumericConvert.h" from inclusions.

**Review by Paul-Licameli — APPROVED** [review]

File: libraries/lib-snapping/SnapUtils.cpp (hunk 1 above)
Comment (crsib) [reply libraries/lib-snapping/SnapUtils.cpp:311]: We have very different view on assertions. There are very little cases where I would agree that assertion without a corresponding `if` is not a programming mistake.

File: src/menus/SelectMenus.cpp (hunk 3 above)
Comment (crsib) [reply src/menus/SelectMenus.cpp:169]: done

File: libraries/lib-snapping/SnapUtils.cpp (hunk 1 above)
Comment (Paul-Licameli) [reply libraries/lib-snapping/SnapUtils.cpp:311]: If that is your opinion, then make the other  constructor consistent.

Comment (Paul-Licameli) [reply libraries/lib-snapping/SnapUtils.cpp:311]: I meant to say, the other unchecked use of mMultiplierFunctor is inconsistent

Comment (crsib) [reply libraries/lib-snapping/SnapUtils.cpp:311]: Great catch, sorry for missing it in the first place

Comment (Paul-Licameli) [reply libraries/lib-snapping/SnapUtils.cpp:311]: I don’t like to review a new assertion that has zero comment with it.  The minimum here would be a ‘@pre’ at the declaration of the constructor.

**Pull request merged**